ambry-store/src/main/java/com/github/ambry/store/Log.java

```diff
@@ -439,14 +451,75 @@ private void free(LogSegment logSegment, boolean isSwapSegment) throws StoreExce
   private void rollOverIfRequired(long writeSize) throws StoreException {
     if (activeSegment.getCapacityInBytes() - activeSegment.getEndOffset() < writeSize) {
       ensureCapacity(writeSize);
-      // this cannot be null since capacity has either been ensured or has thrown.
-      LogSegment nextActiveSegment = segmentsByName.higherEntry(activeSegment.getName()).getValue();
-      logger.info("Rolling over writes to {} from {} on write of data of size {}. End offset was {} and capacity is {}",
-          nextActiveSegment.getName(), activeSegment.getName(), writeSize, activeSegment.getEndOffset(),
-          activeSegment.getCapacityInBytes());
-      activeSegment.dropBufferForAppend();
-      nextActiveSegment.initBufferForAppend();
-      activeSegment = nextActiveSegment;
+      setActiveSegment();
+    }
+  }
+
+  /**
+   * Closes the last log segment periodically if replica is in sealed status.
+   * @param journal The {@link Journal} needs to be clean up after log segment closed.
+   * @throws StoreException if any store exception occurred as part of ensuring capacity.
+   */
+  void autoCloseLastLogSegmentIfQualified(Journal journal) throws StoreException {
+    if (compactionPolicySwitchInfoCounterValueReached()) {
+      //ensure the capacity to open the new log segment and allocate new log segment.
+      //if not able to close the last log segment, continue running the compaction.
+      try {
+        ensureCapacity(0);
+        setActiveSegment();
+        //refresh journal.
+        long startTime = SystemTime.getInstance().milliseconds();
+        if (config.storeSynchronizerInsideWhileLoopEnabled) {
+          journal.cleanUpJournalWithSynchronizeInside(activeSegment.getName());
+        } else {
+          journal.cleanUpJournal(activeSegment.getName());
```

Comment:
Any specific reason that the journal is cleaned up in `Log.java` ?  I feel like it can be done in `PersistentIndex` and thus we don't have to pass the journal variable into this method.

Reply:
Updated to move it to BlobStore.